File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy_price_1 = buy_price_2 = float('inf')
        max_price_1 = max_price_2 = 0
        
        for price in prices:
            buy_price_1 = min(buy_price_1, price)
            max_price_1 = max(max_price_1, price-buy_price_1)
            buy_price_2 = min(buy_price_2, price-max_price_1)
            max_price_2 = max(max_price_2, price-buy_price_2)
            print(buy_price_1, max_price_1, buy_price_2, max_price_2)
        return max_price_2

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.java
class Solution {
    public int maxProfit(int[] prices) {
        int buyPrice1 = Integer.MAX_VALUE;
        int buyPrice2 = Integer.MAX_VALUE;
        int maxPrice1 = 0;
        int maxPrice2 = 0;
        for (int i = 0; i < prices.length; i++) {
            buyPrice1 = Math.min(buyPrice1, prices[i]);
            maxPrice1 = Math.max(maxPrice1, prices[i]-buyPrice1);
            buyPrice2 = Math.min(buyPrice2, prices[i]-maxPrice1);
            maxPrice2 = Math.max(maxPrice2, prices[i]-buyPrice2);
        }
        return maxPrice2;
        
    }
}